File: ReSharper.FSharp/src/Daemon.FSharp/src/Stages/TypeCheckErrorsStage.cs
﻿using JetBrains.Annotations;
using JetBrains.Application.Settings;
using JetBrains.ReSharper.Feature.Services.Daemon;
using JetBrains.ReSharper.Psi;
using JetBrains.ReSharper.Psi.FSharp.Tree;
using JetBrains.Util;
using Microsoft.FSharp.Compiler;

namespace JetBrains.ReSharper.Daemon.FSharp.Stages
{
  [DaemonStage(StagesBefore = new[] {typeof(SetResolvedSymbolsStage)},
    StagesAfter = new[] {typeof(ResolvedIdentifiersHighlightStage)})]
  public class TypeCheckErrorsStage : FSharpDaemonStageBase
  {
    public class TypeCheckErrorsStageProcess : ErrorsStageProcess
    {
      public TypeCheckErrorsStageProcess([NotNull] IDaemonProcess process, [NotNull] FSharpErrorInfo[] errors)
        : base(process, errors)
      {
      }
    }

    protected override IDaemonStageProcess CreateProcess(IFSharpFile fsFile, IDaemonProcess process)
    {
      var errors = fsFile.GetCheckResults(false, process.CreateInterruptChecker())?.Errors;
      return new TypeCheckErrorsStageProcess(process, errors ?? EmptyArray<FSharpErrorInfo>.Instance);
    }

    public override ErrorStripeRequest NeedsErrorStripe(IPsiSourceFile sourceFile,
      IContextBoundSettingsStore settingsStore)
    {
      return ErrorStripeRequest.STRIPE_AND_ERRORS;
    }
  }
}

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/Daemon/Stages/TypeCheckErrorsStage.fs
namespace JetBrains.ReSharper.Plugins.FSharp.Daemon.Stages

open JetBrains.Application.Settings
open JetBrains.ReSharper.Feature.Services.Daemon
open JetBrains.ReSharper.Plugins.FSharp.Daemon.Cs.Stages
open JetBrains.ReSharper.Plugins.FSharp.Psi.Tree
open JetBrains.Util
open Microsoft.FSharp.Compiler.ErrorLogger
open Microsoft.FSharp.Compiler.SourceCodeServices

type TypeCheckErrorsStageProcess(fsFile: IFSharpFile, daemonProcess, logger: ILogger) =
    inherit ErrorsStageProcessBase(fsFile, daemonProcess)

    override x.ShouldAddDiagnostic(error, range) =
        error.Subcategory <> BuildPhaseSubcategory.Parse &&
        base.ShouldAddDiagnostic(error, range)

    override x.Execute(committer) =
        match fsFile.GetParseAndCheckResults(false) with
        | Some results ->
            daemonProcess.CustomData.PutData(FSharpDaemonStageBase.TypeCheckResults, Some results.CheckResults)
            let projectWarnings, fileErrors  =
                results.CheckResults.Errors
                |> Array.partition (fun e ->
                    e.StartLineAlternate = 0 && e.EndLineAlternate = 0 && e.Severity = FSharpErrorSeverity.Warning)

            if not (Array.isEmpty projectWarnings) then
                // https://github.com/Microsoft/visualfsharp/issues/4030
                let errors = Array.fold (fun a e -> a + "\n" + e.ToString()) "" projectWarnings
                logger.LogMessage(LoggingLevel.WARN, "Project warnings during file type check:" + errors)
            x.Execute(fileErrors, committer)
        | _ -> ()


[<DaemonStage(StagesBefore = [| typeof<SyntaxErrorsStage> |], StagesAfter = [| typeof<HighlightIdentifiersStage> |])>]
type TypeCheckErrorsStage(logger: ILogger) =
    inherit FSharpDaemonStageBase()

    override x.CreateProcess(fsFile: IFSharpFile, settings: IContextBoundSettingsStore, daemonProcess: IDaemonProcess) =
        TypeCheckErrorsStageProcess(fsFile, daemonProcess, logger) :> IDaemonStageProcess
